http/src/request/audit_reason.rs

```diff
@@ -0,0 +1,24 @@
+pub trait AuditLogReason {
+    fn reason(self, reason: impl Into<String>) -> Result<Self, AuditLogReasonError>
+    where
+        Self: Sized;
+}
+
+impl AuditLogReasonError {
+    /// The maximum audit log reason length in codepoints.
+    pub const AUDIT_REASON_LENGTH: usize = 512;
+
+    pub(crate) fn validate(reason: String) -> Result<String, AuditLogReasonError> {
+        if reason.chars().count() <= Self::AUDIT_REASON_LENGTH {
+            Ok(reason)
+        } else {
+            Err(AuditLogReasonError::ReasonInvalid { reason })
+        }
+    }
+}
+
+/// The error created when a reason can not be used as configured.
+pub enum AuditLogReasonError {
+    /// Returned when the reason is over 512 UTF-16 characters.
```

Comment:
Does Discord really use UTF-16? If so, I don’t see how the validate function takes that into account.

Reply:
refer to https://github.com/twilight-rs/twilight/pull/146#discussion_r410425722